core/services/gateway/handlers/functions/allowlist/allowlist.go

```diff
@@ -172,17 +208,113 @@ func (a *onchainAllowlist) updateFromContractV1(ctx context.Context, blockNum *b
 	if err != nil {
 		return errors.Wrap(err, "unexpected error during functions_allow_list.NewTermsOfServiceAllowList")
 	}
-	addrList, err := tosContract.GetAllAllowedSenders(&bind.CallOpts{
+
+	var allowedSenderList []common.Address
+	if !a.config.StoreAllowedSendersEnabled {
+		allowedSenderList, err = tosContract.GetAllAllowedSenders(&bind.CallOpts{
+			Pending:     false,
+			BlockNumber: blockNum,
+			Context:     ctx,
+		})
+		if err != nil {
+			return errors.Wrap(err, "error calling GetAllAllowedSenders")
+		}
+	} else {
+		err = a.syncStoredAllowedAndBlockedSenders(ctx, tosContract, blockNum)
+		if err != nil {
+			return errors.Wrap(err, "failed to sync the stored allowed and blocked senders")
+		}
+
+		allowedSenderList, err = a.getAllowedSendersInRange(ctx, tosContract, blockNum)
+		if err != nil {
+			return errors.Wrap(err, "failed to get allowed senders in rage")
+		}
+	}
+
+	a.update(allowedSenderList)
+	return nil
+}
+
+func (a *onchainAllowlist) getAllowedSendersInRange(ctx context.Context, tosContract *functions_allow_list.TermsOfServiceAllowList, blockNum *big.Int) ([]common.Address, error) {
```

Comment:
Putting "InRange" in the name suggests that it takes the range as a parameter. Maybe rename to getAllowedSendersBatched()?